### ISSUE TITLE
GPRO: use_liger_loss + zero3 error

### ISSUE THREAD

**Issue by paul777chen** [huggingface/trl#3368]
### Reproduction

✅ use_liger_loss=true + zero2
✅ use_liger_loss=false + zero3
❌ use_liger_loss=true + zero3

error：size mismatch

### System Info

pytorch=2.6
deepspeed=0.16.4

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by Tavish9** [comment]
@paul777chen could you please try the latest version and report the error trace if something bad happen again?

**Comment by visheratin-pai** [comment]
It seems like the problem is that for zero3, DeepSpeed doesn't fully gather parameter weights, so the layer looks fine but its weights are empty. The problem still exists in 0.20.0.

**Comment by Tavish9** [comment]
@visheratin-pai please show me a smallest example, thanks.

**Comment by visheratin-pai** [comment]
Here is my DS config (from examples):
```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_multinode_launcher: standard
  offload_optimizer_device: none
  offload_param_device: none
  zero3_init_flag: true
  zero3_save_16bit_model: false
  zero_stage: 3
distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
And I run it as usual in multi-note setup (SLURM in my case).

I also just checked that zero2 works. It looks more like accelerate issue than TRL.

**Comment by akshansh47** [comment]
Confirming the bug class. Same root cause as
[transformers#45313](https://github.com/huggingface/transformers/issues/45313):
under ZeRO-3 \`lm_head.weight\` is partitioned to rank-1 shards, and Liger's FLCE
forward reads the weight via attribute access instead of through \`model.forward()\`
— so the gather hook never fires and the size mismatch surfaces.

### User-space stop-gap

Wrap the LM head weight read in \`GatheredParameters\` before Liger's chunked matmul:

\`\`\`python
import deepspeed

with deepspeed.zero.GatheredParameters([head.weight], modifier_rank=None):
    head_weight = head.weight.data.detach().clone()
# pass \`head_weight\` into the Liger loss instead of \`head.weight\`
\`\`\`

Cost is one \`V × H × bf16\` clone per logprob call (~1.9 GB at Qwen3.5 V=152K);
detach is safe because \`lm_head\` is frozen under standard LoRA configs
(\`lora_target_modules\` excludes \`lm_head\`). We've run this in production GRPO
on Qwen3.5-27B for 100+ steps without regression.

### Proper fix

Migrate to FSDP. The TRL Liger integration ([PR #3184](https://github.com/huggingface/trl/pull/3184),
[HF blog](https://huggingface.co/blog/liger-grpo)) reports 40% peak memory
reduction and 1.5–1.8× max batch on FSDP + PEFT. Every production GRPO
framework (verl, SkyRL, LlamaFactory) has converged on FSDP2 for Qwen3-family
training. The \`use_liger_loss + zero3\` path is empirically unsupported.

Happy to PR a \`GatheredParameters\`-wrapped helper in the Liger integration if
maintainers want a stop-gap before users migrate.

**Comment by qgallouedec** [comment]
Hey @paul777chen, sorry for the long delay, catching up on the backlog.

Confirmed: under DeepSpeed ZeRO-3, the Liger path in GRPO reads `unwrapped_model.lm_head.weight` directly (https://github.com/huggingface/trl/blob/51c6d3ca31fb4cc80ff719c0844bbdfcd0feeefb/trl/trainer/grpo_trainer.py#L2330-L2340) without a `GatheredParameters` wrapper, so on partitioned ranks the matmul gets an empty weight and you see the size mismatch. We just landed the same fix pattern for SFT's chunked-NLL loss in f1be82688; we should port `_maybe_gather_lm_head_ctx` to the GRPO Liger path. FYI the flag is now called `use_liger_kernel` (the `use_liger_loss` alias was removed in 0.28.0). Will link the PR here.

cc @kashif
